File: Gold II/4195.cpp
#include <bits/stdc++.h>
using namespace std;

int T, N;
int p[200002], si[200002];
map<string, int> mp;

int fin(int x)
{
    if (p[x] == x)
        return x;
    return p[x] = fin(p[x]);
}

void uni(int a, int b)
{
    if (fin(a) == fin(b))
        return;
    si[fin(a)] += si[fin(b)];
    p[fin(b)] = p[fin(a)];
}

int main()
{
    scanf("%d", &T);
    while (T--)
    {
        scanf("%d", &N);
        mp.clear();
        for (int i = 0, idx = 0; i < N; i++)
        {
            char a[21], b[21];
            scanf("%s %s", a, b);
            string A = a, B = b;
            // printf("%s %s\n", a, b);
            if (mp.find(A) == mp.end())
            {
                mp[A] = idx;
                p[idx] = idx;
                si[idx] = 1;
                idx++;
            }
            if (mp.find(B) == mp.end())
            {

                mp[B] = idx;
                p[idx] = idx;
                si[idx] = 1;
                idx++;
            }

            // printf("%d %d\n", mp[A], mp[B]);
            uni(mp[A], mp[B]);
            printf("%d\n", si[fin(mp[A])]);
        }
    }
}

File: Gold II/4195.py
T = int(input())

def find(a : int, p : list):
    if p[a] == a:
        return a
    p[a] = find(p[a], p)
    return p[a]

def uni(a : int, b : int, p: list):
    p[find(a, p)] = find(b, p)

def size(a : int, p : list):
    result  = 0
    for i in range(len(p)):
        if p[find(a, p)] == p[find(i, p)]:
            result += 1
    return result


for _ in range(T):
    N = int(input())
    names = {}
    p = []
    idx = 0
    for _ in range(N):
        A, B = input().split()
        if A not in names:
            names[A] = idx
            p.append(idx)
            idx += 1
        if B not in names:
            names[B] = idx
            p.append(idx)
            idx += 1
        uni(names[A], names[B], p) 


        print(size(names[A], p))    

